dpytools/http/upload.py

```diff
@@ -0,0 +1,233 @@
+import datetime
+from math import ceil
+import os
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Tuple, Union
+
+from dpytools.http.base import BaseHttpClient
+
+
+class UploadClient(BaseHttpClient):
+    def __init__(self, upload_url: str):
+        # Inherit backoff_max value from BaseHTTPClient.__init__
+        super().__init__()
+        self.upload_url = upload_url
+
+    def upload(
+        self,
+        csv_path: Union[Path, str],
+        s3_bucket: str,
+        florence_access_token: str,
+        chunk_size: int = 5242880,
+    ) -> Tuple[str, str]:
+        """
+        Upload files to the DP Upload Service `upload` endpoint. The file to be uploaded (located at `csv_path`) is chunked (default chunk size 5242880 bytes) and uploaded to an S3 bucket.
+
+        The `s3_bucket` argument should be set as an environment variable and accessed via os.getenv() or similar. `florence_access_token` should be generated via the DP Identity API and passed as an argument to `upload()`.
+
+        Returns the S3 Object key and S3 URL of the uploaded file.
+        """
+        # Create file chunks
+        file_chunks = _create_temp_chunks(csv_path=csv_path, chunk_size=chunk_size)
+
+        # Generate upload request params
+        upload_params = _generate_upload_params(csv_path, chunk_size)
+
+        # Upload file chunks to S3
+        self._upload_file_chunks(file_chunks, upload_params, florence_access_token)
+
+        s3_key = upload_params["resumableIdentifier"]
+        s3_uri = f"s3://{s3_bucket}/{s3_key}"
+
+        # Delete temporary files
+        _delete_temp_chunks(file_chunks)
+        # TODO Replace print statements with logging
+        print("Upload to s3 complete")
+
+        return s3_key, s3_uri
+
+    def upload_new(
+        self,
+        csv_path: Union[Path, str],
+        florence_access_token: str,
+        s3_bucket: str,
+        collection_id: str,
```

Comment:
the swagger spec said collection_id is optional, I'd make it an `Optional[str]` keyword arg then only include it in the params if it's not none.